### ISSUE TITLE
[resolved] OpenID scope customization

### ISSUE THREAD

**Issue by matteo-s** [minio/minio#9238]
It would be advisable to let administrators customize the scopes requested to the IdP.

The new approach (see #9015) is to request all the scopes published by the IdP.
This leads to a number of different problems.

### 1.
This could break secure deployments, since **supported scopes** doesn't equals to **scopes enabled for this client**, and some enforcing authorization servers will deny the request with an ``invalid_scope`` response.

See spec at https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
```
scopes_supported
    RECOMMENDED. JSON array containing a list of the OAuth 2.0 [RFC6749] scope values that this server supports. The server MUST support the openid scope value. Servers MAY choose not to advertise some supported scope values even when this parameter is used, although those defined in [OpenID.Core] SHOULD be listed, if supported. 

```

The mere fact that the server supports a scope doesn't mean that that scope is available for a given client. 
This is a regression bug

### 2
Administrators could choose to request a scope which is not advertised, for a number of reasons such as:

- client-specific 
- sensitive 
- non-standard 
etc.

The spec explicitly supports such solutions.

### 3
This approach leads to an overly-complex user flow, since by OAuth2 spec users are supposed to approve client access to scopes. If Minio requests all the available scopes, users will either need to approve access to all scopes, or filter them if supported by the IdP.
For example, the OpenID Core specification defines (https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims)

* profile
* email
* address
* phone

And many IdP will have additional custom scopes advertised.
Why Minio should request access to scopes it won't use?

For an analogy, it is pretty much the same approach as an Android application requesting *all the permissions* at start.

## Expected Behavior
Don't make out-of-spec assumptions, only ``openid`` is guaranteed to be supported and available.
Let administrators configure additional claims which are available in their environment, if needed.

## Current Behavior
Pre #9015  Minio requests only the standard, always available ``openid`` scope. This is guaranteed to work, but is not configurable by administrators.

After #9015 Minio requests all the scopes advertised by the IdP, resulting in either 
- error in enforcing deployments where some of those scopes are not enabled for the Minio oauth client 
- inability to access custom scopes
- inconveniently complex user-flow with a huge approval step for scopes.

## Possible Solution
As proposed in #9010 let admins configure the scopes, same as other openid properties.

```
export MINIO_IDENTITY_OPENID_SCOPE="openid email groups"
```

Otherwise revert #9015 and request only ``openid``.

While resolving problem (1) may be legitimately not one of your problems, given that is deployment-specific, I think (2) and (3) are far more widespread and thus could qualify for a solution.

## Regression
This is a regression when compared to the previous approach, which requested the always available ``openid`` scope.

**Comment by WolfspiritM** [comment]
I just had a related issue. If there are many (or long) scopes the current implementation also breaks with (in my case) IdentityServer and will result in a "scopes too long" error as it only allowed 300 characters with the default configuration. It'd be really great if one can set the required scopes for minio to get the needed claims and maybe fallback to using all scopes if it is not configured.

**Comment by imartinezortiz** [comment]
I can confirm that this is a problem with Keycloak. I've integrated minio and Keycloak 10.0.2 and If I do not enable all available client scopes for the OIDC client that minio will use , OIDC login does not work.

**Comment by harshavardhana** [comment]
> I can confirm that this is a problem with Keycloak. I've integrated minio and Keycloak 10.0.2 and If I do not enable all available client scopes for the OIDC client that minio will use , OIDC login does not work.

We will add scope customization.. @imartinezortiz @WolfspiritM

**Comment by harshavardhana** [comment]
Closing as fixed and released.

**Comment by WolfspiritM** [comment]
@harshavardhana I gave this a try but it seems like minio browser still uses all scopes? Is that on purpose? Should I open a new issue for that?

Tested with RELEASE.2020-07-02T00-15-09Z

**Comment by imartinezortiz** [comment]
These changes are in master branch but not in release branch, but you can try minio:edge that seems to be build from master branch
